Zeebe 1.2.0-alpha1 can't start on Windows

**Describe the bug**
When starting Zeebe on Windows 10 machine, following exception is thrown:

```
2021-08-05 08:51:30.064 [] [main] ERROR
      io.camunda.zeebe.broker.system - Bootstrap Broker-0 [11/13]: zeebe partitions failed with unexpected exception.
io.camunda.zeebe.journal.JournalException: Failed to flush journal directory after creating segment C:\programs\camunda-cloud-zeebe-1.2.0-alpha1\data\raft-partition\partitions\1\raft-partition-partition-1-1.log
        at io.camunda.zeebe.journal.file.SegmentedJournal.createSegment(SegmentedJournal.java:653) ~[zeebe-journal-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.journal.file.SegmentedJournal.open(SegmentedJournal.java:252) ~[zeebe-journal-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.journal.file.SegmentedJournal.<init>(SegmentedJournal.java:93) ~[zeebe-journal-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.journal.file.SegmentedJournalBuilder.build(SegmentedJournalBuilder.java:133) ~[zeebe-journal-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.atomix.raft.storage.log.RaftLogBuilder.build(RaftLogBuilder.java:123) ~[zeebe-atomix-cluster-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.atomix.raft.storage.RaftStorage.openLog(RaftStorage.java:175) ~[zeebe-atomix-cluster-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.atomix.raft.impl.RaftContext.<init>(RaftContext.java:176) ~[zeebe-atomix-cluster-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:258) ~[zeebe-atomix-cluster-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.atomix.raft.impl.DefaultRaftServer$Builder.build(DefaultRaftServer.java:232) ~[zeebe-atomix-cluster-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.atomix.raft.partition.impl.RaftPartitionServer.buildServer(RaftPartitionServer.java:189) ~[zeebe-atomix-cluster-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.atomix.raft.partition.impl.RaftPartitionServer.initServer(RaftPartitionServer.java:156) ~[zeebe-atomix-cluster-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.atomix.raft.partition.impl.RaftPartitionServer.start(RaftPartitionServer.java:114) ~[zeebe-atomix-cluster-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.atomix.raft.partition.RaftPartition.open(RaftPartition.java:120) ~[zeebe-atomix-cluster-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.atomix.raft.partition.RaftPartitionGroup.lambda$join$6(RaftPartitionGroup.java:204) ~[zeebe-atomix-cluster-1.2.0-alpha1.jar:1.2.0-alpha1]
        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
        at java.util.HashMap$KeySpliterator.forEachRemaining(HashMap.java:1603) ~[?:?]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
        at io.atomix.raft.partition.RaftPartitionGroup.join(RaftPartitionGroup.java:206) ~[zeebe-atomix-cluster-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.atomix.primitive.partition.impl.DefaultPartitionService.start(DefaultPartitionService.java:63) ~[zeebe-atomix-cluster-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.broker.partitioning.PartitionManagerImpl.start(PartitionManagerImpl.java:48) ~[zeebe-broker-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.broker.Broker.partitionsStep(Broker.java:414) ~[zeebe-broker-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.broker.Broker.lambda$initStart$9(Broker.java:258) ~[zeebe-broker-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.broker.bootstrap.StartProcess.lambda$startStepByStep$2(StartProcess.java:64) ~[zeebe-broker-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.broker.bootstrap.StartProcess.takeDuration(StartProcess.java:93) ~[zeebe-broker-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.broker.bootstrap.StartProcess.startStepByStep(StartProcess.java:62) ~[zeebe-broker-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.broker.bootstrap.StartProcess.takeDuration(StartProcess.java:93) ~[zeebe-broker-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.broker.bootstrap.StartProcess.start(StartProcess.java:47) ~[zeebe-broker-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.broker.Broker.internalStart(Broker.java:203) ~[zeebe-broker-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.util.LogUtil.doWithMDC(LogUtil.java:23) [zeebe-util-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.broker.Broker.start(Broker.java:183) [zeebe-broker-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.broker.StandaloneBroker.run(StandaloneBroker.java:60) [camunda-cloud-zeebe-1.2.0-alpha1.jar:1.2.0-alpha1]
        at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:791) [spring-boot-2.5.3.jar:2.5.3]
        at org.springframework.boot.SpringApplication.callRunners(SpringApplication.java:775) [spring-boot-2.5.3.jar:2.5.3]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:345) [spring-boot-2.5.3.jar:2.5.3]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1343) [spring-boot-2.5.3.jar:2.5.3]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1332) [spring-boot-2.5.3.jar:2.5.3]
        at io.camunda.zeebe.broker.StandaloneBroker.main(StandaloneBroker.java:47) [camunda-cloud-zeebe-1.2.0-alpha1.jar:1.2.0-alpha1]
Caused by: java.nio.file.AccessDeniedException: C:\programs\camunda-cloud-zeebe-1.2.0-alpha1\data\raft-partition\partitions\1
        at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:89) ~[?:?]
        at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103) ~[?:?]
        at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108) ~[?:?]
        at sun.nio.fs.WindowsFileSystemProvider.newFileChannel(WindowsFileSystemProvider.java:116) ~[?:?]
        at java.nio.channels.FileChannel.open(FileChannel.java:292) ~[?:?]
        at java.nio.channels.FileChannel.open(FileChannel.java:345) ~[?:?]
        at io.camunda.zeebe.util.FileUtil.flush(FileUtil.java:46) ~[zeebe-util-1.2.0-alpha1.jar:1.2.0-alpha1]
        at io.camunda.zeebe.journal.file.SegmentedJournal.createSegment(SegmentedJournal.java:650) ~[zeebe-journal-1.2.0-alpha1.jar:1.2.0-alpha1]
```

**To Reproduce**
<!--
Steps to reproduce the behavior

If possible add a minimal reproducer code sample
 - when using the Java client: https://github.com/zeebe-io/zeebe-test-template-java

-->

**Expected behavior**
<!-- A clear and concise description of what you expected to happen. -->

**Log/Stacktrace**
<!-- If possible add the full stacktrace or Zeebe log which contains the issue. -->

<details><summary>Full Stacktrace</summary>
 <p>

```
<STACKTRACE>
```

</p>
</details>


**Environment:**
 - OS: Win 10
 - Zeebe Version: 1.2.0-alpha1
 - Configuration: Elasticsearch exporter

--- a/.github/workflows/os-smoke-test.yml
+++ b/.github/workflows/os-smoke-test.yml
@@ -1,0 +1,34 @@
+name: OS Support Smoke Test
+
+on:
+  push:
+    branches: [ develop ]
+  pull_request:
+    branches: [ develop ]
+
+jobs:
+  build:
+    name: Run smoke tests on ${{ matrix.os }}
+    timeout-minutes: 10
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ macos-latest, windows-latest, ubuntu-latest ]
+
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up JDK 11
+        uses: actions/setup-java@v2
+        with:
+          java-version: '11'
+          distribution: 'adopt'
+      - name: Cache Maven packages
+        uses: actions/cache@v2.1.6
+        with:
+          path: ~/.m2
+          key: ${{ runner.os }}-m2-${{ hashFiles('**/pom.xml') }}
+          restore-keys: ${{ runner.os }}-m2
+      - name: Build with Maven
+        run: mvn -B -pl qa/integration-tests -am install -DskipTests -DskipChecks
+      - name: Run smoke test
+        run: mvn -B -pl qa/integration-tests verify -P smoke-test -DskipUTs -DskipChecks

--- a/journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournal.java
+++ b/journal/src/main/java/io/camunda/zeebe/journal/file/SegmentedJournal.java
@@ -675,7 +675,7 @@ public final class SegmentedJournal implements Journal {
     // necessary to flush the directory to ensure that the file itself is visible as an entry of
     // that directory after recovery
     try {
-      FileUtil.flush(directory.toPath());
+      FileUtil.flushDirectory(directory.toPath());
     } catch (final IOException e) {
       throw new JournalException(
           String.format("Failed to flush journal directory after creating segment %s", segmentFile),

--- a/qa/integration-tests/pom.xml
+++ b/qa/integration-tests/pom.xml
@@ -273,5 +273,28 @@
       <scope>test</scope>
     </dependency>
   </dependencies>
+
+  <profiles>
+    <!--
+      this profile will run all tests annotated with @SmokeTest. the profile is mainly used to
+      ensure that we can run basic tests on Windows, macOS, and Linux, thereby ensuring minimal
+      multi-platform support for development.
+      -->
+    <profile>
+      <id>smoke-test</id>
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-failsafe-plugin</artifactId>
+            <configuration>
+              <failIfNoTests>false</failIfNoTests>
+              <groups>smoke-test</groups>
+            </configuration>
+          </plugin>
+        </plugins>
+      </build>
+    </profile>
+  </profiles>
 </project>
 

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/smoke/SmokeTest.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/smoke/SmokeTest.java
@@ -1,0 +1,21 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.it.smoke;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+
+@Target({ElementType.TYPE, ElementType.METHOD})
+@Retention(RetentionPolicy.RUNTIME)
+@Tag("smoke-test")
+@Test
+public @interface SmokeTest {}

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/smoke/StandaloneBrokerIT.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/smoke/StandaloneBrokerIT.java
@@ -23,7 +23,6 @@ import io.restassured.http.ContentType;
 import io.restassured.specification.RequestSpecification;
 import java.util.Objects;
 import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
@@ -49,7 +48,7 @@ final class StandaloneBrokerIT {
   }
 
   /** A simple smoke test to ensure the broker starts and can perform basic functionality. */
-  @Test
+  @SmokeTest
   void smokeTest() {
     // given
     final var processId = Strings.newRandomValidBpmnId();

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/smoke/StandaloneGatewayIT.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/smoke/StandaloneGatewayIT.java
@@ -46,7 +46,7 @@ final class StandaloneGatewayIT {
   private GatewayCfg config;
 
   /** A simple smoke test which checks that the gateway can start and accept requests. */
-  @Test
+  @SmokeTest
   void smokeTest() {
     // given
     try (final var client =

--- a/util/pom.xml
+++ b/util/pom.xml
@@ -16,7 +16,6 @@
     <dependency>
       <groupId>org.agrona</groupId>
       <artifactId>agrona</artifactId>
-      <scope>provided</scope>
     </dependency>
 
     <dependency>
@@ -38,10 +37,12 @@
       <groupId>com.fasterxml.jackson.core</groupId>
       <artifactId>jackson-databind</artifactId>
     </dependency>
+
     <dependency>
       <groupId>com.fasterxml.jackson.core</groupId>
       <artifactId>jackson-core</artifactId>
     </dependency>
+
     <dependency>
       <groupId>com.fasterxml.jackson.core</groupId>
       <artifactId>jackson-annotations</artifactId>

--- a/util/src/main/java/io/camunda/zeebe/util/FileUtil.java
+++ b/util/src/main/java/io/camunda/zeebe/util/FileUtil.java
@@ -25,6 +25,7 @@ import java.nio.file.SimpleFileVisitor;
 import java.nio.file.StandardOpenOption;
 import java.nio.file.attribute.BasicFileAttributes;
 import java.util.Objects;
+import org.agrona.SystemUtil;
 import org.slf4j.Logger;
 
 public final class FileUtil {
@@ -34,10 +35,9 @@ public final class FileUtil {
 
   /**
    * Ensures updates to the given path are flushed to disk, with the same guarantees as {@link
-   * FileChannel#force(boolean)}. This can be used for both directory and files. Note that if you
-   * already have a file channel open, then use the {@link FileChannel#force(boolean)} method
-   * directly. This method is mostly useful when flushing directories after file creation and/or
-   * file rename operations.
+   * FileChannel#force(boolean)}. This can be used for files only; for directories, call {@link
+   * #flushDirectory(Path)}. Note that if you already have a file channel open, then use the {@link
+   * FileChannel#force(boolean)} method directly.
    *
    * @param path the path to synchronize
    * @throws IOException can be thrown on opening and on flushing the file
@@ -46,6 +46,27 @@ public final class FileUtil {
     try (final var channel = FileChannel.open(path, StandardOpenOption.READ)) {
       channel.force(true);
     }
+  }
+
+  /**
+   * This method flushes the given directory. Note that on Windows this is a no-op, as Windows does
+   * not allow to flush directories except when opening the sys handle with a specific backup flag,
+   * which is not possible to do without custom JNI code. However, as the main use case here is for
+   * durability to sync the parent directory after creating a new file or renaming it, this is safe
+   * to do as Windows (or rather NTFS) does not need this.
+   *
+   * @param path the path to synchronize
+   * @throws IOException can be thrown on opening and on flushing the file
+   */
+  public static void flushDirectory(final Path path) throws IOException {
+    // Windows does not allow flushing a directory except under very specific conditions which are
+    // not possible to produce with the standard JDK; it's also not necessary to flush a directory
+    // in Windows
+    if (SystemUtil.isWindows()) {
+      return;
+    }
+
+    flush(path);
   }
 
   /**
@@ -61,7 +82,7 @@ public final class FileUtil {
   public static void moveDurably(final Path source, final Path target, final CopyOption... options)
       throws IOException {
     Files.move(source, target, options);
-    flush(target.getParent());
+    flushDirectory(target.getParent());
   }
 
   public static void deleteFolder(final String path) throws IOException {

